Cache/Hashing.lean

```diff
@@ -11,8 +11,27 @@ namespace Cache.Hashing
 
 open System IO
 
-/-- We cache the hash of each file for faster lookup -/
-abbrev HashM := StateT IO.HashMap IO
+structure HashMemo where
+  depsMap : Lean.HashMap FilePath (Array FilePath)
+  hashMap : HashMap
+  deriving Inhabited
+
+partial def insertDeps (hashMap : HashMap) (path : FilePath) (hashMemo : HashMemo) : HashMap :=
+  if hashMap.contains path then hashMap else
+  match (hashMemo.depsMap.find? path, hashMemo.hashMap.find? path) with
+  | (some deps, some hash) => deps.foldl (insertDeps · · hashMemo) (hashMap.insert path hash)
+  | _ => hashMap
+
+def HashMemo.filterByPatterns (hashMemo : HashMemo) (patterns : List String) : IO HashMap := do
+  let mut hashMap := default
+  for pattern in patterns do
+    if hashMemo.hashMap.contains pattern then
+      hashMap := insertDeps hashMap pattern hashMemo
+    else IO.println s!"No match for {pattern}"
```

Comment:
Does this mean "no cache was avaialble" or "the lean file you asked for doesn't even exist"? I think the second should just be an error.

Reply:
The second option. Why an error? Skipping is a better UX so the user doesn't have to re-type what he/she has already got right

Reply:
It's not how command line tools usually work. For instance `cat file_that_doesnt_exist` raises an exit code.

> so the user doesn't have to re-type what he/she has already got right

This is what shell history is for